### PR TITLE
[663] agent topics

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

- Separated agent mapper interface topics and c8ytopics.
- Agent interface topics are in `agent_interface` crate.
- Moved download code that was remaining inside the agent to download crate
- renamed sm_json crate to agent_interface.
- updated agent and mapper code accordingly.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: crates/core/c8y_smartrest/Cargo.toml [hunk 1]
```diff
@@ -5,11 +5,11 @@ authors = ["thin-edge.io team <info@thin-edge.io>"]
 edition = "2018"
 
 [dependencies]
+agent_mapper_interface = { path = "../agent_mapper_interface" }
```
Comment (didier-wenzek) [inline crates/core/c8y_smartrest/Cargo.toml:8]: This dependency is really suspicious. A crate that is related to SmartRest should be unrelated to the agent.

* This is due to the `SmartRestSerializerError::UnsupportedOperationStatus` enum.
* This error case has to be replaced with `UnsupportedOperationStatus { status: String }`.

File: crates/core/agent_mapper_interface/src/topic.rs [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+    #[error("Topic {topic} is unknown.")]
+    UnknownTopic { topic: String },
+}
+
 #[derive(Debug, Clone, PartialEq)]
-pub(crate) enum IncomingTopic {
+pub enum IncomingTopic {
     SoftwareListResponse,
     SoftwareUpdateResponse,
     SmartRestRequest,
```
Comment (didier-wenzek) [inline crates/core/agent_mapper_interface/src/topic.rs:14]: The `SmartRestRequest` topic is unrelated to the agent and must be removed from here.

File: crates/core/agent_mapper_interface/src/topic.rs [hunk 3]
```diff
@@ -1,17 +1,22 @@
...
 
+#[derive(thiserror::Error, Debug)]
+pub enum TopicError {
+    #[error("Topic {topic} is unknown.")]
+    UnknownTopic { topic: String },
+}
+
 #[derive(Debug, Clone, PartialEq)]
-pub(crate) enum IncomingTopic {
+pub enum IncomingTopic {
```
Comment (didier-wenzek) [inline crates/core/agent_mapper_interface/src/topic.rs:11]: The name is misleading after the move from the mapper to this agent interface.

```suggestion
pub enum ResponseTopic {
```

File: crates/core/agent_mapper_interface/src/topic.rs [hunk 4]
```diff
@@ -22,47 +27,47 @@ impl IncomingTopic {
...
+    type Error = TopicError;
 
     fn try_from(value: Topic) -> Result<Self, Self::Error> {
         value.name.try_into()
     }
 }
 
 #[derive(Debug, Clone, PartialEq)]
-pub(crate) enum OutgoingTopic {
+pub enum OutgoingTopic {
```
Comment (didier-wenzek) [inline crates/core/agent_mapper_interface/src/topic.rs:62]: The name is misleading after the move from the mapper to this agent interface.

```suggestion
pub enum RequestTopic {
```

File: crates/core/agent_mapper_interface/src/topic.rs [hunk 5]
```diff
@@ -22,47 +27,47 @@ impl IncomingTopic {
...
     SoftwareListRequest,
     SoftwareUpdateRequest,
     SmartRestResponse,
```
Comment (didier-wenzek) [inline crates/core/agent_mapper_interface/src/topic.rs:65]: This topic is unrelated to the agent and must be defined by the mapper.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
Need to be fixed:
* The smartrest crate must not depend on the agent interface.
* The agent interface must not include topics that are related to Cumulocity.

**Review by rina23q — COMMENTED** [review]
I have just one comment.

File: crates/core/agent_mapper_interface/src/topic.rs [hunk 6]
```diff
@@ -1,17 +1,22 @@
-use crate::sm_c8y_mapper::error::MapperTopicError;
 use mqtt_client::{MqttClientError, Topic};
 use std::convert::{TryFrom, TryInto};
 
+#[derive(thiserror::Error, Debug)]
+pub enum TopicError {
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/agent_mapper_interface/src/topic.rs]: moved

File: crates/core/agent_mapper_interface/src/topic.rs (hunk 4 above)
Comment (PradeepKiruvale) [reply crates/core/agent_mapper_interface/src/topic.rs]: done

File: crates/core/agent_mapper_interface/src/topic.rs (hunk 3 above)
Comment (PradeepKiruvale) [reply crates/core/agent_mapper_interface/src/topic.rs]: done

File: crates/core/c8y_smartrest/Cargo.toml (hunk 1 above)
Comment (PradeepKiruvale) [reply crates/core/c8y_smartrest/Cargo.toml]: it also requires even OperationStatus

File: crates/common/download/Cargo.toml [hunk 7]
```diff
@@ -7,8 +7,8 @@ license = "Apache-2.0"
 description = "download_manager"
 
 [dependencies]
+agent_mapper_interface = { path = "../../core/agent_mapper_interface" }
 backoff = { version = "0.3", features = ["tokio"] }
-json_sm = { path = "../../core/json_sm" }
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/common/download/Cargo.toml]: Moved download related code into download crate

File: crates/core/agent_mapper_interface/Cargo.toml [hunk 8]
```diff
@@ -1,15 +1,16 @@
 [package]
-name = "json_sm"
+name = "agent_mapper_interface"
 version = "0.5.0"
 authors = ["thin-edge.io team <info@thin-edge.io>"]
 edition = "2018"
 license = "Apache-2.0"
-description = "json_sm defines the thin-edge json extension for software management"
+description = "agent_mapper_interface defines the mqtt topics and thin-edge json extension for software management and restart operations "
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/agent_mapper_interface/Cargo.toml]: fixed

File: crates/core/agent_mapper_interface/src/topic.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/core/agent_mapper_interface/src/topic.rs]: separated the agent_interface topics and c8y topics.

File: crates/core/agent_interface/Cargo.toml [hunk 9]
```diff
@@ -1,15 +1,17 @@
...
 license = "Apache-2.0"
-description = "json_sm defines the thin-edge json extension for software management"
+description = "agent_interface defines all the mqtt topics and messages to be used to invoke the agent."
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
 csv = "1.1"
+download = { path = "../../common/download" }
+mqtt_client = { path = "../../common/mqtt_client" }
```
Comment (didier-wenzek) [inline crates/core/agent_interface/Cargo.toml:14]: Not a request for change, just a question. I wonder if having a dependency to `mqtt_client` is a good idea, knowing that this is just to use `Topic` vs `String` for the topics.

* Would the change to unchecked `String` topics implies a lot of refactoring ?

File: crates/core/c8y_smartrest/Cargo.toml [hunk 10]
```diff
@@ -5,11 +5,12 @@ authors = ["thin-edge.io team <info@thin-edge.io>"]
 edition = "2018"
 
 [dependencies]
+agent_interface = { path = "../agent_interface" }
```
Comment (didier-wenzek) [inline crates/core/c8y_smartrest/Cargo.toml:8]: I really don't see why a Cumulocity specific crate has to depend on the agent interface.

* What would be the impact of removing this dependency?

File: crates/core/tedge_mapper/Cargo.toml [hunk 11]
```diff
@@ -33,9 +34,9 @@ c8y_translator = { path = "../c8y_translator" }
 chrono = "0.4"
 clock = { path = "../../common/clock" }
 csv = "1.1"
+download = { path = "../../common/download" }
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/Cargo.toml:37]: I'm surprised to see `download` as a *new* dependency, since the mapper is actually doing no download.

* We have a `pub use download::*;` in the `agent_interface`. That should be enough to manipulate the download infos.

File: crates/core/tedge_mapper/src/sm_c8y_mapper/error.rs [hunk 12]
```diff
@@ -1,21 +1,16 @@
...
-}
-
 #[derive(thiserror::Error, Debug)]
 pub enum SMCumulocityMapperError {
     #[error("Invalid MQTT Message.")]
     InvalidMqttMessage,
 
     #[error(transparent)]
-    InvalidTopicError(#[from] MapperTopicError),
+    InvalidTopicError(#[from] TopicError),
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/sm_c8y_mapper/error.rs:10]: I would rename the error type and make explicit that this is related to the agent (minor).

``` suggestion
    InvalidTopicError(#[from] agent_interface::UnknowTopic),
```

File: crates/core/tedge_mapper/src/sm_c8y_mapper/mapper.rs [hunk 13]
```diff
@@ -143,26 +145,27 @@ where
             debug!("Topic {:?}", message.topic.name);
             debug!("Mapping {:?}", message.payload_str());
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/sm_c8y_mapper/mapper.rs:146]: Please, remove all these `debug!` messages.

File: crates/core/tedge_mapper/src/sm_c8y_mapper/tests.rs [hunk 14]
```diff
@@ -11,7 +11,7 @@ use std::{io::Write, time::Duration};
 use tedge_config::{ConfigRepository, TEdgeConfig, TEdgeConfigLocation};
 use tokio::task::JoinHandle;
 
-const TEST_TIMEOUT_MS: Duration = Duration::from_millis(1000);
+const TEST_TIMEOUT_MS: Duration = Duration::from_millis(3000);
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/sm_c8y_mapper/tests.rs:14]: Why this has to be changed?

File: crates/core/tedge_mapper/src/sm_c8y_mapper/topic.rs [hunk 15]
```diff
@@ -1,83 +1,99 @@
...
-            Self::SoftwareUpdateRequest => r#"tedge/commands/req/software/update"#,
-            Self::SmartRestResponse => r#"c8y/s/us"#,
-            Self::RestartRequest => r#"tedge/commands/req/control/restart"#,
+impl TryFrom<String> for MapperSubscribeTopic {
+    type Error = TopicError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match value.as_str() {
+            r#"c8y/s/ds"# => Ok(MapperSubscribeTopic::SmartRestRequest),
+            r#"tedge/commands/res/software/list"# => Ok(MapperSubscribeTopic::ResponseTopic(
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/sm_c8y_mapper/topic.rs:68]: It's s a good point to have a enum that encapsulates all the input topic of the mapper
However, the current implementation of the `try_from` method defeats the main point of this PR
that is to encapsulate the agent's topic names inside the agent interface crate.

The `MapperSubscribeTopic ::try_from` implementation must use the `agent_interface::ResponseTopic::try_from` method as well as the SmartRestRequest::try_from` method.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
* The `MapperSubscribeTopic::try_from` method must be rewritten.
* I wonder if the code can be simplified using `String` instead `Topic`  in the `agent_interface` crate.

File: crates/core/tedge_mapper/src/sm_c8y_mapper/tests.rs (hunk 14 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/sm_c8y_mapper/tests.rs:14]: Reverted back

File: crates/core/tedge_mapper/Cargo.toml (hunk 11 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/Cargo.toml:37]: Mapper uses the REST APIs to send the software list operation result. The URL creation and authorization for REST APIs uses some code from the download crate, that's the reason why the mapper still needs the download crate. The code related to the URL can be moved to a new URL crate.

File: crates/core/c8y_smartrest/Cargo.toml (hunk 10 above)
Comment (PradeepKiruvale) [reply crates/core/c8y_smartrest/Cargo.toml:8]: The SoftwareUpdateResponse and OperationStatus are defined in the agent_interface crate, these are used in  Cumulocity specific crate. Need some more refactoring is required to move this common code to some new crate.

File: crates/core/tedge_mapper/src/sm_c8y_mapper/mapper.rs (hunk 13 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/sm_c8y_mapper/mapper.rs:146]: removed

File: crates/core/agent_interface/Cargo.toml (hunk 9 above)
Comment (PradeepKiruvale) [reply crates/core/agent_interface/Cargo.toml:14]: To create `mqtt topics`  from the `Strings topics` the mqtt_client crate is used here and used in mapper and some tests in agent-interface crate. To remove this the mqtt topic creation can be moved into the mapper directly.

**Review by didier-wenzek — APPROVED** [review]
Approved
